R/key_colnames.R

```diff
@@ -1,19 +1,20 @@
 #' @export
 key_colnames.recipe <- function(x, ...) {
-  possible_keys <- c("geo_value", "time_value", "key")
-  keys <- x$var_info$variable[x$var_info$role %in% possible_keys]
-  keys[order(match(keys, possible_keys))] %||% character(0L)
+  geo_key <- x$var_info$variable[x$var_info$role %in% "geo_value"]
+  time_key <- x$var_info$variable[x$var_info$role %in% "time_value"]
+  keys <- x$var_info$variable[x$var_info$role %in% "key"]
+  c(geo_key, keys, time_key) %||% character(0L)
 }
 
 #' @export
 key_colnames.epi_workflow <- function(x, ...) {
   # safer to look at the mold than the preprocessor
   mold <- hardhat::extract_mold(x)
-  possible_keys <- c("geo_value", "time_value", "key")
   molded_names <- names(mold$extras$roles)
-  keys <- map(mold$extras$roles[molded_names %in% possible_keys], names)
-  keys <- unname(unlist(keys))
-  keys[order(match(keys, possible_keys))] %||% character(0L)
+  geo_key <- names(mold$extras$roles[molded_names %in% "geo_value"]$geo_value)
+  time_key <- names(mold$extras$roles[molded_names %in% "time_value"]$time_value)
+  keys <- names(mold$extras$roles[molded_names %in% "key"]$key)
+  c(geo_key, keys, time_key) %||% character(0L)
```

Comment:
By same as above, why not keep 
```r
possible_keys <- c("geo_value", "key", "time_value")
```
then the rest can be unchanged?

Reply:
Same as the other key_colnames function, I rewrote in a way that made sense to me. This function in particular was difficult for me to follow since we're having to index into mold$extras$roles and navigate these awkward nested tidymodels containers.

Reply:
Ok, yes, lets revert if it still works.

Reply:
Hm, unfortunately `possible_keys <- c("geo_value", "key", "time_value")` doesn't work. The gist is that `match(keys, possible_keys)` returns NA for keys that aren't in possible_keys, those are exactly the other_keys, and order puts NAs last by default (otherwise it puts them in the front). You can play with it here if you'd like:

```r
library(epipredict)
library(testthat)
library(tidyverse)

key_colnames_test <- function(x, ...) {
  possible_keys <- c("geo_value", "key", "time_value")
  keys <- x$var_info$variable[x$var_info$role %in% possible_keys]
  keys[order(match(keys, possible_keys))] %||% character(0L)
}

my_data <- tibble::tibble(
  geo_value = rep(c("ca", "fl", "pa"), each = 3),
  time_value = rep(seq(as.Date("2020-06-01"), as.Date("2020-06-03"),
    by = "day"
  ), length.out = length(geo_value)),
  pol = rep(c("blue", "swing", "swing"), each = 3), # extra key
  state = rep(c("ca", "fl", "pa"), each = 3), # extra key
  value = 1:length(geo_value) + 0.01 * rnorm(length(geo_value))
) %>%
  as_epi_df(
    other_keys = c("state", "pol")
  )

my_recipe <- epi_recipe(my_data) %>%
  step_epi_ahead(value, ahead = 7) %>%
  step_epi_naomit()

# order of the additional keys may be different
expect_equal(key_colnames_test(my_recipe), c("geo_value", "state", "pol", "time_value"))

possible_keys <- c("geo_value", "key", "time_value")
keys <- my_recipe$var_info$variable[my_recipe$var_info$role %in% possible_keys]
keys
match(keys, possible_keys)
order(match(keys, possible_keys))
```

Can also just go with my code and rewrite it later if you'd like. My code passes unit tests, so it should be functionally the same.